discojs/src/aggregator/aggregator.ts

```diff
@@ -62,24 +62,53 @@ export abstract class Base<T> extends EventEmitter<{'aggregation': T }> {
     this.contributions = Map()
     this._nodes = Set()
 
-    // On every aggregation, update the object's state to match the current aggregation
-    // and communication rounds.
-    this.on('aggregation', () => this.nextRound())
+    // On each aggregation, increment
+    // updates the aggregator's state to proceed to the next communication round.
+    // If all communication rounds were performed, proceeds to the next aggregation round
+    // and empties the collection of stored contributions.
+    this.on('aggregation', () => {
+      if (++this._communicationRound === this.communicationRounds) {
+        this._communicationRound = 0
+        this._round++
+        this.contributions = Map()
+      }
+    })
   }
 
   /**
    * Adds a node's contribution to the aggregator for the given aggregation and communication rounds.
    * The aggregation round is increased whenever a new global model is obtained and local models are updated.
    * Within one aggregation round there may be multiple communication rounds (such as for the decentralized secure aggregation
-   *  which requires multiple steps to obtain a global model)
+   * which requires multiple steps to obtain a global model)
    * The contribution will be aggregated during the next aggregation step.
    * @param nodeId The node's id
    * @param contribution The node's contribution
-   * @param round aggregation round of the contribution was made
    * @param communicationRound communication round the contribution was made within the aggregation round
-   * @returns boolean, true if the contribution has been successfully taken into account or False if it has been rejected
+   * @returns undefined, if the contribution is invalid, o.w. a promise for the weight update
+   */
+  abstract add(nodeId: client.NodeID, contribution: WeightsContainer, communicationRound?: number): Promise<WeightsContainer>
+  
+  /**
+   * Create a promise which resolves when enough contributions are received and 
+   * local updates are aggregated. 
+   * If the aggregator has enough contribution then we can aggregate the weights
+   * directly (and emit the 'aggregation' event)
+   * Otherwise we wait for the 'aggregation' event which will be emitted once 
+   * enough contributions are received
+   * 
+   * @returns a promise for the aggregated weights
    */
-  abstract add (nodeId: client.NodeID, contribution: T, round: number, communicationRound?: number): boolean
+  protected createAggregationPromise(): Promise<WeightsContainer> {
+    if (this.isFull()) {
+      const aggregatedWeights = this.aggregate()
+      // Emitting the 'aggregation' communicates the aggregation to other clients and
+      // takes care of incrementing the round
+      this.emit('aggregation', aggregatedWeights)
+      return Promise.resolve(aggregatedWeights);
+    }
+    // Wait for the aggregation event to be emitted
+    return new Promise<WeightsContainer>((resolve) => this.once('aggregation', resolve))
```

Comment:
this is effectively equivalent, but a bit more readable (single exit).

```suggestion
    // Wait for the aggregation event to be emitted
    const ret = new Promise<WeightsContainer>((resolve) => this.once('aggregation', resolve));
    
    if (this.isFull()) {
      const aggregatedWeights = this.aggregate()
      // Emitting the 'aggregation' communicates the aggregation to other clients and
      // takes care of incrementing the round
      this.emit('aggregation', aggregatedWeights)
    }
    
    return ret
```

but I'm reading a bit more and looks like it's only when exiting the `add` method. consider splitting the `add` call, one public just a now but not abstract, one protected abstract (smth like `_add`) which does the internal work. the public `add` will then ensure that it uses `createAggregationPromise` (and potentially log before calling the internal `_add`)